### PR TITLE
Core package

### PR DESCRIPTION
**Pull request by dboslee**
The core package implements the following

- JobStore for storing and getting jobs by ID
- Job which provides a way to start, stop, get the status, and stream the output of a job
- Stream which allows for many channels to subscribe and receive published data. This will be used to stream logs to the client

### CODE REVIEW

File: pkg/core/stream.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	defer s.mu.Unlock()
+	for i, sub := range s.subs {
+		if ch != sub {
+			continue
+		}
+		// Reslice and close
+		s.subs = append(s.subs[:i], s.subs[i+1:]...)
+		close(ch)
+	}
+}
```
Comment (wadells) [inline pkg/core/stream.go:74]: The behavior when a single channel is subscribed multiple times is odd -- one Unsubscribe will remove all the subscriptions -- and panic on the 2nd close.  In a more fleshed out project we'd want to avoid one channel being subscribed multiple times or fix the behavior here to remove one at a time. I'd say these concerns are largely beyond the scope of this project -- a note about intended behavior would suffice.

File: pkg/core/stream.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+// Pipe publishes bytes to the stream from a reader
+func (s *Stream) Pipe(r io.Reader) {
+	for {
+		b := make([]byte, 1024)
+		n, err := r.Read(b)
+
+		// Close the stream on EOF
+		if err != nil {
```
Comment (wadells) [inline pkg/core/stream.go:97]: Could `err != nil && err != io.EOF`? If yes, would that info be useful to maintainers for debugging?

File: pkg/core/stream.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	if s.closed {
+		return
+	}
+	s.closed = true
+
+	// Clean up subs
+	for _, ch := range s.subs {
+		close(ch)
+	}
+	s.subs = make([]chan []byte, 1)
```
Comment (wadells) [inline pkg/core/stream.go:118]: Can you add a Publish after close test case?  I *think* line 118 + no closed short circuit in publish sets up for a send to a nil channel.

**Review by wadells — COMMENTED** [review]
Looks quite clean.  A word or two about why you wrote stream.go over something like  https://golang.org/pkg/io/#MultiReader and https://golang.org/pkg/io/#MultiWriter wouldn't go awry.

File: pkg/core/job.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	return j.Cmd.Process.Signal(os.Kill)
+}
+
+// Start runs a job and handles errors
+func (j *Job) Start() error {
+	done := make(chan error)
+	go func() {
+		done <- j.run()
+	}()
+	err := <-done
```
[the remark this answers is not in the document]
Comment (dboslee) [reply pkg/core/job.go:83]: Yeah not sure what I was thinking. I will change this.

File: pkg/core/stream.go (hunk 2 above)
Comment (dboslee) [reply pkg/core/stream.go:97]: Good point. I will log it in that case.

File: pkg/core/stream.go (hunk 1 above)
Comment (dboslee) [reply pkg/core/stream.go:74]: I had not considered a channel subscribing more than once. Seems easy enough to add a check in Subscribe for this

File: pkg/core/stream.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
+package core
+
+import (
+	"io"
+	"sync"
+)
+
+// MergeStreams sends the combined output of two streams to a given channel
+func MergeStreams(ch chan []byte, s1, s2 *Stream) {
```
Comment (r0mant) [inline pkg/core/stream.go:10]: I think the way logs handling is currently implement is quite a bit more complicated than it has to be TBH.

If you intend to store all process outputs in memory, wouldn't something like a synchronized buffer be much simpler? Wouldn't need all this subscribing/publishing/merging machinery.

That said, current approach still stores everything in memory as I as understand. Think the design can be tweaked a little bit to be more accommodating to processes that produce large amounts of logs such as that running `yes` wouldn't crash the server with OOM for example?

File: pkg/core/stream.go (hunk 3 above)
Comment (dboslee) [reply pkg/core/stream.go]: > Looks quite clean. A word or two about why you wrote stream.go over something like https://golang.org/pkg/io/#MultiReader and https://golang.org/pkg/io/#MultiWriter wouldn't go awry.

I could have used MultiReader to merge the output but stream.go gives me a way to dynamically add subscribers to the output stream which will be useful for sending output to the client.

File: pkg/core/stream.go (hunk 5 above)
Comment (dboslee) [reply pkg/core/stream.go:10]: To address the potential memory issue and simplify things I can make the following changes. 

Scrap the stream solution and instead go with writing to a temporary directory that can be cleaned up before the server exits.

Then add a naive tail function for the client to use that will read and retry periodically when it hits EOF until the job exits or hits some other error. An alternative to this would be to implement something using https://github.com/fsnotify/fsnotify or use https://github.com/hpcloud/tail but for the scope of this project I think it is overkill to implement a proper solution and pull in these dependencies.

**Comment by dboslee** [comment]
> Looks like a good start and good to go once Roman's comment is addressed. How will job status be handled? Will this simply be constructed from `ProcessState` when queried?

Yes I have the following methods for Job that will be used. Exited() along with ExitCode() and Error() if applicable

File: pkg/core/stream.go [hunk 6]
```diff
@@ -95,6 +104,9 @@ func (s *Stream) Pipe(r io.Reader) {
 
 		// Close the stream on EOF
 		if err != nil {
+			if err != io.EOF {
+				fmt.Println("unexpected error", err)
```
Comment (wadells) [inline pkg/core/stream.go:108]: fmt.Println isn't concurrency safe -- at least if you expect output not to be garbled. Also I wouldn't necessarily recommend raw printing to stdout from a library function.  If you want to keep this, can you remark on how you'd change your approach in a production system?  If this code is removed in the refactor -- no need to address this further.

File: pkg/core/stream_test.go [hunk 7]
```diff
@@ -55,3 +55,9 @@ func TestClose(t *testing.T) {
 		t.Errorf("expected channel to be closed")
 	}
 }
+
+func TestPublishAfterClose(t *testing.T) {
+	stream := core.NewStream()
+	stream.Close()
+	stream.Publish(want)
```
Comment (wadells) [inline pkg/core/stream_test.go:62]: There are no subscribers in this test case -- would this have caught the nil channel send?

File: pkg/core/stream.go (hunk 6 above)
Comment (dboslee) [reply pkg/core/stream.go:108]: Yeah this will probably get taken out. But for any future log statements I will use the log package over fmt.

**Review by knisbet — APPROVED** [review]
I don't really have anything additional to the others comments. Looking forward to seeing how this continues.

File: pkg/core/stream.go (hunk 5 above)
Comment (r0mant) [reply pkg/core/stream.go]: @dboslee Sorry I didn't see this reply earlier for some reason... Yeah, writing logs to files sounds good.

Comment (dboslee) [reply pkg/core/stream.go]: @r0mant no worries. Just added these changes.

**Comment by dboslee** [comment]
Forgot to mention the reasoning behind adding status now. After @bernardjkim brought it up I thought about it more and decided it would be a log simpler this way over having to check multiple things like if job.Exited() && job.Error(). Also avoids any potential sync issues there.

File: pkg/core/output.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+type OutputBuffer struct {
+	f *os.File
+}
+
+// NewOutputBuffer creates a OutputBuffer instance
+func NewOutputBuffer() (*OutputBuffer, error) {
+
+	// TODO: Swap out tempfile for a permanent solution
+	// This is easier to setup/cleanup
+	f, err := ioutil.TempFile("", "job-worker-output-*")
```
Comment (r0mant) [inline pkg/core/output.go:19]: More of a nit, but for an even simpler approach I'd probably just use `ioutil.TempDir()` so each job just gets its own temp. directory and you wouldn't need to bother opening/closing files here.

File: pkg/core/job.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		return err
+	}
+	j.UpdateStatus(Running)
+
+	err = j.writeOutput(output)
+
+	// If we hit an error after the job starts we can kill and
+	if err != nil {
+		j.Kill()
```
Comment (r0mant) [inline pkg/core/job.go:141]: Not sure how I feel about this design decision TBH. I.e. why would we terminate a job if we failed to write its logs?

File: pkg/core/job.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+// UpdateStatus updates a jobs err
+func (j *Job) UpdateStatus(status JobStatus) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+	j.status = status
+}
+
+// Interrupt sends a SIGINT to the process
+func (j *Job) Interrupt() error {
+	return j.Cmd.Process.Signal(os.Interrupt)
```
Comment (r0mant) [inline pkg/core/job.go:94]: What happens to the job status in this case and in case of Kill below?

**Review by r0mant — COMMENTED** [review]
A few more comments but feel free to address them in follow-up PRs so we can keep moving forward.

**Review by bernardjkim — APPROVED** [review]

File: pkg/core/job.go (hunk 9 above)
Comment (dboslee) [reply pkg/core/job.go:141]: Agreed probably better to have no logs and let the job complete. A counter argument would be if the output for the job is actually important to the user. But even then it may be better to let the job complete but have some status message that the logs failed. Anyways I will just log the error here.

File: pkg/core/job.go (hunk 10 above)
Comment (dboslee) [reply pkg/core/job.go:94]: It should get set to error since the cmd.Wait should returns an error

File: pkg/core/job.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+	w, err := j.OutputBuf.NewWriter()
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+	b := make([]byte, 1024)
+
+	// Loop until io.EOF or other error
+	for {
```
[the remark this answers is not in the document]
Comment (dboslee) [reply pkg/core/job.go:160]: Looks like it will. I will swap those out in the next PR
